### PR TITLE
bpo-32355: Optimize asyncio.gather()

### PR DESCRIPTION
**Pull request by 1st1**
I also believe the this PR makes `asyncio.gather()` code simpler (besides also making it 10-15% faster).

https://bugs.python.org/issue32355

### CODE REVIEW

File: Lib/asyncio/tasks.py [hunk 1]
```diff
@@ -605,56 +604,68 @@ def gather(*coros_or_futures, loop=None, return_exceptions=False):
...
-    for i, fut in enumerate(children):
-        fut.add_done_callback(functools.partial(_done_callback, i))
+    arg_to_fut = {}
+    children = []
+    nfuts = 0
+    nfinished = 0
+    for arg in coros_or_futures:
+        if arg not in arg_to_fut:
+            fut = ensure_future(arg, loop=loop)
+            if loop is None:
```
Comment (LiraNuna) [inline Lib/asyncio/tasks.py:657]: It seems like the check for multiple loops (see line 622) has been removed. Is this intentional?
If it is, you may want to add a test for that.

Comment (1st1) [reply Lib/asyncio/tasks.py:657]: ensure_future does it. No need to do it twice.

Comment (LiraNuna) [reply Lib/asyncio/tasks.py:657]: `ensure_future` validates that the passed in `loop` argument matches that of the future passed in. If I were to pass 3 futures each having their own different loop, a new `_GatheringFuture` will be created tied to the loop that was passed to `gather` (which can be a separate loop as well).

Comment (LiraNuna) [reply Lib/asyncio/tasks.py:657]: I think I understand what you mean by that, `ensure_future` will make sure that the future/coro passed is bound to the same loop - either current or what was passed in. Thanks for explaining.

File: Lib/asyncio/tasks.py [hunk 2]
```diff
@@ -605,56 +604,68 @@ def gather(*coros_or_futures, loop=None, return_exceptions=False):
...
+
+        if nfinished == nfuts:
+            # All futures are done; create a list of results
+            # and set it to the 'outer' future.
+            results = []
+
+            for fut in children:
+                if fut.cancelled():
+                    res = futures.CancelledError()
+                elif fut._exception is not None:
```
Comment (asvetlov) [inline Lib/asyncio/tasks.py]: `_exception` and `_result` is not a part of Future public API

File: Lib/asyncio/tasks.py [hunk 3]
```diff
@@ -605,56 +604,68 @@ def gather(*coros_or_futures, loop=None, return_exceptions=False):
...
                 # Mark exception retrieved.
                 fut.exception()
             return
 
-        if fut.cancelled():
-            res = futures.CancelledError()
-            if not return_exceptions:
-                outer.set_exception(res)
+        if not return_exceptions:
+            if fut.cancelled():
```
Comment (asvetlov) [inline Lib/asyncio/tasks.py]: Both checks can be merged:
```
exc = fut.exception()
if exc is not None:
    outer.set_exception(exc)
```
it handles `CancelledError` too.

File: Lib/asyncio/tasks.py [hunk 4]
```diff
@@ -605,56 +604,68 @@ def gather(*coros_or_futures, loop=None, return_exceptions=False):
...
+        if arg not in arg_to_fut:
+            fut = ensure_future(arg, loop=loop)
+            if loop is None:
+                loop = fut._loop
+            if fut is not arg:
+                # 'arg' was not a Future, therefore, 'fut' is a new
+                # Future created specifically for 'arg'.  Since the caller
+                # can't control it, disable the "destroy pending task"
+                # warning.
+                fut._log_destroy_pending = False
```
Comment (asvetlov) [inline Lib/asyncio/tasks.py:664]: Unrelated question: maybe add `log_destroy_pending` keyword-only parameter?
Often I want to see cancelled coroutines in logs.

File: Lib/asyncio/tasks.py (hunk 2 above)
Comment (1st1) [reply Lib/asyncio/tasks.py]: well, old code used it, i'm not adding anything new

File: Lib/asyncio/tasks.py (hunk 3 above)
Comment (1st1) [reply Lib/asyncio/tasks.py]: no, it will raise CancelledError, not return it. We can't merge these checks.

File: Lib/asyncio/tasks.py (hunk 2 above)
Comment (1st1) [reply Lib/asyncio/tasks.py]: I thought about this for a while.  Let's use only public Future API in asyncio.  I've updated this PR and will make a new one to fix all remaining places in asyncio.

Comment (1st1) [reply Lib/asyncio/tasks.py]: Actually, it's just one other place where we access `Future._exception` or `Future._result` in asyncio -- it's `base_events.py`.  I've fixed it as part of this PR.

**Review by asvetlov — APPROVED** [review]

**Pull request merged**